### PR TITLE
BasicConstraint CSR support

### PR DESCRIPTION
**Pull request by tam7t**
Addresses https://github.com/cloudflare/cfssl/issues/495

This adds more complete support for the `BasicConstraint` extension when generating and singing CSRs. Continues from the work in https://github.com/cloudflare/cfssl/pull/576 adding...
- `pathlen` to the CSRs generated by cfssl
- use the `pathlen` info from CSRs when signing requests

### CODE REVIEW

**Comment by tam7t** [comment]
@lziest would appreciate a look

**Comment by lziest** [comment]
Yes, will definitely review it tomorrow.

File: signer/local/testdata/inter_pathlen_0-key.pem [hunk 1]
```diff
@@ -0,0 +1,5 @@
+-----BEGIN EC PRIVATE KEY-----
```
Comment (lziest) [inline signer/local/testdata/inter_pathlen_0-key.pem]: I realize you would never use the -key and cert.pem in those local signer tests, better not check them in. Just keep the generated around.

**Comment by lziest** [comment]
I see the approach here is to let CSR contains CA pathlen info and pass it down to the final cert. The other side of control is the CA signing config, where we define how signer controls what the final certificate should look like. We still want the signing config take priority over CSR value, i.e. signing config should check/override CSR.

Can you simplify the CSR part of the code to have appendCAInfoToCSR only append unspecified pathlen, and let signer to decide what value should be filled in?

**Comment by benwhite-deltas** [comment]
Thanks for movement on this. If I may comment from a user perspective:

> We still want the signing config take priority over CSR value

I think this is good, provided the signing config pathlen value is lower than the CSR pathlen value - ultimately I believe the lower of the two should be used if they are both specified, as the signing authority shouldn't generate a certificate with a pathlen longer than that being requested.

Additionally it would be useful to reflect back to the user a warning if the pathlen value requested in the CSR has been lowered by the signing value.

**Comment by tam7t** [comment]
I think that the CSR should contain the requested `pathlen` so that cfssl generated CSRs would be compatible with other certificate authorities (in case I wanted to generate an intermediate and leaf certs with cfssl, but have it signed by my my organization's active directory CA).

I can add a check/warning that the requested cert has a lower pathlen than the signing cert (if the singing cert has a pathlen restriction).

File: signer/local/testdata/inter_pathlen_0-key.pem (hunk 1 above)
Comment (tam7t) [inline signer/local/testdata/inter_pathlen_0-key.pem]: fixed!

File: csr/csr.go [hunk 2]
```diff
@@ -176,6 +177,12 @@ func (cr *CertificateRequest) Name() pkix.Name {
 	return name
 }
 
+// BasicConstraints CSR information RFC 5280, 4.2.1.9
+type BasicConstraints struct {
```
Comment (lziest) [inline csr/csr.go]: do you think this can be an unexported struct?

**Comment by lziest** [comment]
rebase on top of current master may help me see the diff better

**Comment by lziest** [comment]
I agree we should make CSR more complete/compatible so it works better with other existing CAs.

**Comment by tam7t** [comment]
This PR is pretty large, I'm thinking I will split it into ~3 smaller PR's that build off eachother and should hopefully be easier to review.

**Comment by kisom** [comment]
@tam7t That would help out a lot --- I've had a hard time finding time to sit down and go through this.

**Pull request closed without merging**